### PR TITLE
Add operator== to Quaternion

### PR DESCRIPTION
**Pull request by eoineoineoin**
Adds an operator== to Quaternion so that two instances of quaternions can be evaluated for equality.

I didn't see a unit test for this comparable functionality in Vector3, so assumed it's trivial enough, but can add one if desired.

### CODE REVIEW

**Comment by eoineoineoin** [comment]
Not sure why the CLA test is failing; maybe an email address mismatch? I've filled in the CLA with my eoin.mcloughlin@havok.com Google account, which should be associated with this Github username.

**Comment by rakudrama** [comment]
I believe having `operator==` and `hashCode` for these classes is a mistake.

They should not have been added to the VectorN classes. They should be deprecated and removed.

Floating point arithmetic involves rounding errors which means the same vector/quaternion/whatever calculated in slightly different ways that are mathematically equivalent might not compare as equal.  __(a+b)+c__ is often not the same value as __a+(b+c)__ once rounding errors happen.

The correct way to compare values that are computed by a process that introduces (small) errors is to check if they are within some tolerance. Exactly how one does this depends on the application - but `q1.relativeError(q2) < 1e-6` might be a starting point.

The presence of `operator==` leads developers to make unfounded assumptions that comparing these values is useful. We get questions all the time about why[ `0.1 + 0.2 != 0.3`](https://dartpad.dev/?id=4d6861744f78fc34caf97b686a2c10a0). Having `operator==` just leads to more subtle versions of the same problem.

Given these standard problems with floating-point arithmetic, I believe it is a disservice to allow these values to be used as hash-table keys.

**Comment by eoineoineoin** [comment]
Hi Stephen,

I think you've conflated two orthogonal issues here; I agree that (unfortunately,) not everyone understands IEEE754, there are valid reasons for comparing exact values of two floating point numbers - case in point, the unit tests in this repository.

Removing operator== doesn't solve the user education issue - a user who's confused by a failure of `expect(vector1, vector2)` will be just as confused when `expect(vector1.storage, vector2.storage)` and `expect(vector1.x, vector2.x)` fails. Removing the operator== only makes it harder for people implementing those valid use-cases.

The language _does_ allow equality comparisons between floating point numbers, so it's pretty strange that higher dimensional embeddings of floats wouldn't support the same feature.

Eoin

File: lib/src/vector_math/quaternion.dart [hunk 1]
```diff
@@ -394,6 +394,18 @@ class Quaternion {
         w * ow - x * ox - y * oy - z * oz);
   }
 
+  /// Check if two quaternions are the same.
+  @override
+  bool operator ==(Object other) =>
+      (other is Quaternion) &&
+      (_qStorage[0] == other._qStorage[0]) &&
```
Comment (rakudrama) [inline lib/src/vector_math/quaternion.dart:401]: Compare in decreasing order of index - this helps array bounds check elimination (see `operator *` above)

**Review by rakudrama — CHANGES_REQUESTED** [review]
Make sure to update the _64 version (run tool/generate_vector_math_64.dart)

**Comment by eoineoineoin** [comment]
Thanks for the feedback, Stephen. Have applied your suggestions. I note that running generate_vector_math_64.dart, it generated a diff which I didn't apply - looks like a bug in that generation code? It seems to have replaced a doc string incorrectly.

```
diff --git a/lib/src/vector_math_64/noise.dart b/lib/src/vector_math_64/noise.dart
index 701ba4a..2dd17c7 100644
--- a/lib/src/vector_math_64/noise.dart
+++ b/lib/src/vector_math_64/noise.dart
@@ -25,8 +25,9 @@ part of '../../vector_math_64.dart';
  * This is based on the implementation of Simplex Noise by Stefan Gustavson
  * found at: http://webstaff.itn.liu.se/~stegu/simplexnoise/SimplexNoise.java
  */
+
 @Deprecated('This API will be removed '
-    '(see https:github.com/google/vector_math.dart/issues/270)')
+    '(see https:github.com/google/vector_math_64.dart/issues/270)')
 class SimplexNoise {
   static final List<List<double>> _grad3 = <List<double>>[
     <double>[1.0, 1.0, 0.0],
```

**Comment by eoineoineoin** [comment]
CLA issue where it depends on the email address got me again. Have force-pushed an update with matching address.

**Review by rakudrama — APPROVED** [review]

**Pull request merged**